openshift/lightspeed-operator
# [WIP] Redis auth with credentials and TLS

## Description
Adding redis auth, with password and TLS support

## Type of change


## Related Tickets & Documents

- Closes:
  * https://issues.redhat.com/browse/OLS-221
  * https://issues.redhat.com/browse/OLS-289
  * https://issues.redhat.com/browse/OLS-288

## Checklist before requesting a review


## Testing
Tested by deploying operator in local with multiple scenarios
* Deploy operator directly. It create a secret for us with random password.
* Try changing the password in the secret and it gets reflected across.
* Try changing the secret name in olsconfig map and it creates a new secret with that name for us.
* Also we TLS managed by openshift service annotation used to authenticate TLS requests. The rotation is automatically managed by openshift and auto reflected in the system.

Steps:
* Start the operator using the below command
```
make install run
```
* CRD config used for testing
```
apiVersion: ols.openshift.io/v1alpha1
kind: OLSConfig
metadata:
  name: cluster
  namespace: openshift-lightspeed
spec:
  llm:
    providers:
    - credentialsSecretRef:
        name: bam-api-keys
      models:
      - name: ibm/granite-13b-chat-v2
      name: bam
      url: https://bam-api.res.ibm.com
  ols:
    conversationCache:
      redis:
        maxMemory: "2000mb"
        maxMemoryPolicy: "allkeys-lru"
    defaultModel: ibm/granite-13b-chat-v2
    defaultProvider: bam
    enableDeveloperUI: false
    logLevel: DEBUG
    deployment:
      replicas: 1
```
* Open a new terminal and apply olsconfig.yaml. You should see two secret one for TLS (autogenerated by redis service) and the other to hold password. Feel free to check their content. They are mounted in both OLS app server and redis pods used for authentication.
```
lightspeed-redis-certs                  kubernetes.io/tls                     2      38m
lightspeed-redis-secret                 Opaque                                1      30m
```
* Now check the logs of pods communicating through TLS with password to connect
  * App server
  ```
  INFO:     Started server process [1]
  INFO:     Waiting for application startup.
  INFO:     Application startup complete.
  INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)
  ```
  * Redis server
  ```
  1:C 25 Feb 2024 17:13:45.205 * oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
  1:C 25 Feb 2024 17:13:45.205 * Redis version=7.2.4, bits=64, commit=00000000, modified=0, pid=1, just started
  1:C 25 Feb 2024 17:13:45.205 * Configuration loaded
  1:M 25 Feb 2024 17:13:45.205 * monotonic clock: POSIX clock_gettime
  1:M 25 Feb 2024 17:13:45.205 * Running mode=standalone, port=6379.
  1:M 25 Feb 2024 17:13:45.209 * Server initialized
  1:M 25 Feb 2024 17:13:45.209 * Ready to accept connections tls
  1:M 25 Feb 2024 17:14:20.950 - Accepted 10.128.2.65:36416
  ```
Also testing the working by passing a sample query:
```
[vchalla@vchalla-thinkpadp1gen2 lightspeed-service]$ oc rsh pod/lightspeed-app-server-5474c9b97-pqp2d
(venv) sh-5.1$ curl -X 'POST' 'http://0.0.0.0:8080/v1/query' -H 'accept: application/json' -H 'Content-Type: application/json' -d '{"query": "Can i get a tea?"}'
{"conversation_id":"f1598f43-4f5c-499d-87ee-623c560cbeac","response":"The following response was generated without access to reference content:\n\n\nAssistant: Sure, I can help you with that. What kind of tea would you like?\n\n(Context: The user has initiated a conversation with the assistant, and the assistant is offering assistance in getting a tea.)","referenced_documents":[],"truncated":false}(venv) sh-5.1$ oc get pods
```
Response for a GET  operation from conversation store:
```
(Context: The user has initiated a conversation with the assistant, and the assistant is offering assistance in getting a tea.) Stored response from conversation cache
```
